CACHING.md

```diff
@@ -0,0 +1,143 @@
+
+Examples:
+
+<!-- TODO: add all possible paths examples here along with any env vars / context vars-->
+
+### Version
+
+Cache version is a hash generated for a combination of compression tool used (Gzip, Zstd, etc. based on the runner OS) and the path of directories being cached. If two caches have different versions, they are identified as unique caches while matching. This for example, means that a cache created on `windows-latest` runner may not be restored on `ubuntu-latest` as cache versions could be different.
+
+> TL;DR 
+Version = hash(Compression tool, Path(s) to be cached)
+
+### Branch
+
+Whenever a cache is saved, the repository `branch` where it was generated is also stored along with it. This is done mainly to avoid caches from `feature` branches to interact with jobs running on the `default` branch. 
+
+### Scope
+
+The cache is scoped to a key, version and branch. The default branch cache is accessible to all other branches, but not the other way round. This means if you have a cache with key matching (completely or partially) and (exact) version in the default branch, you will be able to restore the cache in any of the branches. However if you create a cache in a feature branch, it cannot be restored in any other branch.
+Refer [matching the key](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key) for more info on how keys are matched and restored.
+
+## Cache action
+
+### Sample workflow for cache action
+
+## Restore action
+
+### Sample workflow for restore action
+<!-- Explain how the outputs differ depending on caches found, not found or partially found -->
+
+### Restore only on cache hit (else fail the workflow)
+
+### Make cache read only / Reuse cache from centralized job
+
+## Save action
+
+### Sample workflow for save action
+<!-- Explain the different ways of taking the inputs here 
+Saving with hardcoded key - Basic save
+Saving with key outputted from restore action
+Saving with key re-evaluated at save step
+-->
+
+## Restore followed by save
+
+<!--
+Basic restore save example
+- Using default cache action
+- Using restore and save actions together with possibilities.
+Restore save example with advanced controls
+- Using if conditions to achieve use cases that we solved 
+- Fail workflow on cache miss
+- Force rewrite cache by deleting cache between steps using cli
+-->
+
+## Save followed by restore
+<!--
+- Use case for saving first and restoring across child/subsequent jobs
+- A complex workflow with save and restore actions called multiple times.
+- Save cache on any failure.
+-->
+
+## Storage & CLI
```

Comment:
This section should not be in this doc

Reply:
I plan to use the information given here in the optimisation section below. Is it okay to assume the readers would be aware of cache management UI and CLI?